File: src/main/java/nl/wernerkroneman/Drawy/ModelEditor/KnowledgeModelInterpreter.java
package nl.wernerkroneman.Drawy.ModelEditor;

import nl.wernerkroneman.Drawy.Modelling.Model;
import nl.wernerkroneman.Drawy.ParseTreeMatcher.PhraseTree;

import java.util.Map;
import java.util.Stack;

/**
 * Generate a creation command for the object(s) described in this sentence part.* @param scene A supplier that provides a scene when executed in which to create the object
 */
public class KnowledgeModelInterpreter
        implements MainInterpreter.InterpretedObjectFactory<Model> {

    MainInterpreter subtreeInterpreter;
    private Knowledge knowledge;

    public KnowledgeModelInterpreter(MainInterpreter subtreeInterpreter,
                                     Knowledge knowledge) {
        this.subtreeInterpreter = subtreeInterpreter;
        this.knowledge = knowledge;
    }

    @Override
    public Class getInterpretedTypePrediction() {
        return Model.class;
    }

    @Override
    public Model createObject(Map<String, PhraseTree> capturings) {

        // Try retrieveing the object by name
        String objectName = capturings.get("object name").getRootWord();
        Model model = knowledge.getObject(objectName);

        return model;
    }
}

File: src/main/java/nl/wernerkroneman/Drawy/ModelEditor/Interpreters/KnowledgeModelInterpreter.kt
package nl.wernerkroneman.Drawy.ModelEditor.Interpreters

import nl.wernerkroneman.Drawy.ModelEditor.Knowledge
import nl.wernerkroneman.Drawy.ModelEditor.MainInterpreter
import nl.wernerkroneman.Drawy.Modelling.Model
import nl.wernerkroneman.Drawy.ParseTreeMatcher.PhraseTree
import java.util.Stack

/**
 * Interpreter for Model.
 */
class KnowledgeModelInterpreter(internal var subtreeInterpreter: MainInterpreter,
                                private val knowledge: Knowledge) : MainInterpreter.InterpretedObjectFactory<Model> {

    override val interpretedTypePrediction: Class<*>
        get() = Model::class.java

    override fun createObject(capturings: Map<String, PhraseTree>): Model? {

        // Try retrieving the object by name
        val objectName = capturings["object name"]!!.getRootWord()
        val model = knowledge.getObject(objectName)

        return model
    }
}
